// @steamship/cli/dist/commands/plugin/instance/list.js
import { CliUx, Command, Flags } from '@oclif/core';
import { PluginInstance } from '@steamship/client';
import { getAuthedClient } from '../../../helpers/get-client.js';
import { showTable } from '../../../helpers/show-table.js';
import { getConfiguration } from '../../../wizards/get-config.js';
import { getPlugin } from '../../../wizards/get-plugin.js';
function toList(d) {
    return [d.handle || '', d.description || '', d.createdAt || ''];
}
export default class List extends Command {
    static description = 'Lists the instances of an Plugin Version.';
    static examples = [
        'ship plugin:instance:list',
        'ship plugin:instance:list --plugin <pluginHandle>'
    ];
    static flags = {
        plugin: Flags.string({ char: 'a', description: 'the Plugin handle' })
    };
    static args = [];
    async run() {
        const { args, flags } = await this.parse(List);
        const client = await getAuthedClient();
        const config = await getConfiguration({});
        const plugin = await getPlugin({ client, config, providedHandle: flags.plugin });
        const request = await PluginInstance.list(client, { pluginId: plugin.id });
        if (!request.output) {
            CliUx.ux.error('Could not retrieve version list from Steamship.');
        }
        const instances = request.output?.pluginInstances;
        if (!instances || instances.length == 0) {
            CliUx.ux.info("You don't have any plugin instances yet.");
            CliUx.ux.info('Create one with:   ship plugin:instance:create');
            CliUx.ux.exit();
        }
        showTable({
            title: `Instances of ${plugin.handle}`,
            headers: ['Handle', 'Description', 'Created At'],
            data: instances.map((instance) => toList(instance))
        });
    }
}
