// @steamship/cli/dist/commands/project/bundle.js
import { CliUx, Command, Flags } from '@oclif/core';
import { bundlePackage } from '../../helpers/bundle-package.js';
import { getProjectConfigWizard } from '../../wizards/get-project-config.js';
export default class Bundle extends Command {
    static description = 'Bundles a deployment package (for testing)';
    static examples = ['ship util:bundle'];
    static flags = {
        help: Flags.help()
    };
    static args = [];
    async run() {
        const { args, flags } = await this.parse(Bundle);
        const config = await getProjectConfigWizard({
            checkName: true,
            checkHandle: true,
            checkVersion: true,
            checkType: true,
            checkEntrypoint: true,
            checkPluginParams: true,
            willDeploy: true
        });
        if (config.type != 'plugin' && config.type != 'package') {
            CliUx.ux.error('The util:bundle command must be run.js from the root folder of a Steamship Plugin or Package project.');
        }
        CliUx.ux.log(`Bundling ${config.handle}, version ${config.version}.zip`);
        let zipLocation = null;
        try {
            zipLocation = await bundlePackage();
        }
        catch (e) {
            CliUx.ux.error(e);
        }
        CliUx.ux.log(`Wrote bundle to ${zipLocation}`);
    }
}
