// @steamship/cli/dist/helpers/package-types.js
import { readFileSync } from 'fs';
import chalk from 'chalk';
export var PackageType;
(function (PackageType) {
    PackageType[PackageType["Jupyter"] = 0] = "Jupyter";
    PackageType[PackageType["Python"] = 1] = "Python";
    PackageType[PackageType["Typescript"] = 2] = "Typescript";
    PackageType[PackageType["React"] = 3] = "React";
    PackageType[PackageType["Unknown"] = 4] = "Unknown";
})(PackageType || (PackageType = {}));
export function getPackageType(rootPath) {
    let packageJson = null;
    try {
        const rawFile = readFileSync(`${rootPath}/package.json`, 'utf-8');
        packageJson = JSON.parse(rawFile);
    }
    catch {
        console.error(`${chalk.red('Error. ')} Unable to read package.json from new project: ${rootPath}/package.json`);
        return PackageType.Unknown;
    }
    if (packageJson == null) {
        console.error(`${chalk.red('Error. ')} Unable to read package.json from new project: ${rootPath}/package.json`);
        return PackageType.Unknown;
    }
    if (typeof packageJson.steamship == 'undefined') {
        console.error(`${chalk.red('Error. ')} Example project's package.json did not include a 'steamship' setting block.`);
        return PackageType.Unknown;
    }
    if (typeof packageJson.steamship.packageType == 'string') {
        console.error(`${chalk.red('Error. ')} Example project's package.json did not include a 'packageType' string parameter in the 'steamship' setting block.`);
        return PackageType.Unknown;
    }
    switch (packageJson.steamship.packageType.toLowerCase()) {
        case 'jupyter':
            return PackageType.Jupyter;
        case 'python':
            return PackageType.Python;
        case 'typescript':
            return PackageType.Typescript;
        case 'react':
            return PackageType.React;
        default:
            return PackageType.Unknown;
    }
}
function displayPostInstallInstructionsJupyter() {
    console.log('TODO: Display instructions');
}
function displayPostInstallInstructionsUnknown() {
}
export function displayPostInstallInstructions(appName, appPath, packageType) {
    console.log(`${chalk.green('Success!')} Created ${appName} at ${appPath}`);
    switch (packageType) {
        case PackageType.Jupyter:
            displayPostInstallInstructionsJupyter();
        default:
            displayPostInstallInstructionsUnknown();
    }
    console.log();
}
