// @steamship/cli/dist/commands/package/list.js
import { CliUx, Command } from '@oclif/core';
import { Package } from '@steamship/client';
import { getAuthedClient } from '../../helpers/get-client.js';
import { showTable } from '../../helpers/show-table.js';
function toList(d) {
    return [d.handle || '', d.description || '', d.createdAt || ''];
}
export default class List extends Command {
    static description = 'Lists your apps.';
    static examples = ['ship package:list'];
    static flags = {};
    static args = [];
    async run() {
        const client = await getAuthedClient();
        const request = await Package.list(client);
        if (!request.output) {
            CliUx.ux.error('Could not retrieve package list from Steamship.');
        }
        const apps = request.output?.packages;
        if (!apps || apps.length == 0) {
            CliUx.ux.info("You don't have any apps yet.");
            CliUx.ux.info('Create one with:      ship create <name>');
            CliUx.ux.info('Then deploy it with:  ship deploy');
            CliUx.ux.exit();
        }
        showTable({
            title: 'Packages',
            headers: ['Handle', 'Description', 'Created At'],
            data: request.output.packages.map(toList)
        });
    }
}
