// @steamship/cli/dist/wizards/get-plugin-version.js
import { CliUx } from '@oclif/core';
import { PluginVersion } from '@steamship/client';
import { getHandle } from './get-handle.js';
export async function getPluginVersion(params) {
    if (!params.plugin) {
        CliUx.ux.warn('Unable to load plugin instance - plugin missing.');
        CliUx.ux.exit(-1);
    }
    const handle = await getHandle({
        provided: params.providedHandle,
        config: params.config,
        type: 'version'
    });
    const versionR = await PluginVersion.get(params.client, { pluginId: params.plugin.id, handle });
    if (!versionR.output) {
        CliUx.ux.warn('Unable to load plugin version.');
        CliUx.ux.exit(-1);
    }
    return versionR.output;
}
const CREATE_FAIL = 'Unable to create new PluginVersion.';
export async function createPluginVersion(params) {
    let version;
    if (!params.plugin?.id) {
        CliUx.ux.warn(`${CREATE_FAIL} Missing Package ID.`);
        process.exit(0);
    }
    if (!params.deploymentZipLocation) {
        CliUx.ux.warn(`${CREATE_FAIL} Missing deployment ZIP location.`);
        process.exit(0);
    }
    try {
        const createParams = {
            pluginId: params.plugin.id,
            filename: params.deploymentZipLocation,
            handle: params.handle,
            configTemplate: params.configTemplate
        };
        version = await PluginVersion.create(params.client, createParams);
    }
    catch (e) {
        CliUx.ux.warn(`${CREATE_FAIL} Empty PluginVersion result from server. Have you incremented or changed your plugin version? \n ${e}`);
        process.exit(0);
    }
    try {
        await version?.wait({ maxTimeoutSeconds: 300 });
    }
    catch (e) {
        CliUx.ux.warn(`${CREATE_FAIL} Deployment failed. \n ${e.statusMessage ?? e.message ?? e}`);
        process.exit(0);
    }
    if (version?.failed()) {
        CliUx.ux.warn(`${CREATE_FAIL} Deployment failed. \n ${version?.statusMessage}`);
        process.exit(0);
    }
    if (!version.output) {
        CliUx.ux.warn(`${CREATE_FAIL} Deployment failed. Empty Plugin Version response on task.`);
        process.exit(0);
    }
    return version.output;
}
