// @steamship/client/build/main/lib/log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogEntry = void 0;
exports["default"] = getLogger;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _tslog = require("tslog");
var _utils = require("./utils.js");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function getLogger(name) {
  if ((0, _utils.isNode)()) {
    var log = new _tslog.Logger({
      minLevel: 'warn',
      // WARN
      name: name
    });
    return log;
  } else {
    var _log = new _tslog.Logger({
      suppressStdOutput: true,
      minLevel: 'warn',
      // WARN
      type: 'json'
    });
    return _log;
  }
}
var LogEntry = /*#__PURE__*/function () {
  function LogEntry() {
    (0, _classCallCheck2["default"])(this, LogEntry);
  }
  (0, _createClass2["default"])(LogEntry, null, [{
    key: "list",
    value: function () {
      var _list = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, params, config) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return client.post('logs/list', _objectSpread({}, params), _objectSpread({}, config));
            case 2:
              return _context.abrupt("return", _context.sent);
            case 3:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      function list(_x, _x2, _x3) {
        return _list.apply(this, arguments);
      }
      return list;
    }()
  }]);
  return LogEntry;
}();
exports.LogEntry = LogEntry;
