// @steamship/client/build/main/lib/package.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Package = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _EXPECT = function _EXPECT(client, data) {
  return new Package(client, data);
};
var _EXPECT_LIST = function _EXPECT_LIST(client, data) {
  return {
    packages: data.packages.map(function (x) {
      return _EXPECT(client, x);
    })
  };
};
var Package = /*#__PURE__*/function () {
  function Package(client, params) {
    (0, _classCallCheck2["default"])(this, Package);
    (0, _defineProperty2["default"])(this, "id", void 0);
    (0, _defineProperty2["default"])(this, "handle", void 0);
    (0, _defineProperty2["default"])(this, "description", void 0);
    (0, _defineProperty2["default"])(this, "profile", void 0);
    (0, _defineProperty2["default"])(this, "readme", void 0);
    (0, _defineProperty2["default"])(this, "isPublic", void 0);
    (0, _defineProperty2["default"])(this, "userId", void 0);
    (0, _defineProperty2["default"])(this, "userHandle", void 0);
    (0, _defineProperty2["default"])(this, "createdAt", void 0);
    (0, _defineProperty2["default"])(this, "updatedAt", void 0);
    (0, _defineProperty2["default"])(this, "featured", void 0);
    (0, _defineProperty2["default"])(this, "client", void 0);
    this.client = client;
    this.id = params.id;
    this.handle = params.handle;
    this.userId = params.userId;
    this.createdAt = params.createdAt;
    this.updatedAt = params.updatedAt;
    this.isPublic = params.isPublic;
    this.description = params.description;
    this.readme = params.readme;
    this.profile = params.profile;
    this.featured = params.featured;
  }
  (0, _createClass2["default"])(Package, [{
    key: "update",
    value: function () {
      var _update = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(config) {
        var params;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              params = {
                id: this.id,
                handle: this.handle,
                readme: this.readme,
                profile: this.profile,
                description: this.description
              };
              _context.next = 3;
              return this.client.post('package/update', _objectSpread({}, params), _objectSpread({
                expect: _EXPECT,
                responsePath: 'package'
              }, config));
            case 3:
              return _context.abrupt("return", _context.sent);
            case 4:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function update(_x) {
        return _update.apply(this, arguments);
      }
      return update;
    }()
  }], [{
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, params, config) {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return client.post('package/create', _objectSpread({}, params), _objectSpread(_objectSpread({}, config), {}, {
                expect: _EXPECT,
                responsePath: 'package'
              }));
            case 2:
              return _context2.abrupt("return", _context2.sent);
            case 3:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      function create(_x2, _x3, _x4) {
        return _create.apply(this, arguments);
      }
      return create;
    }()
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, params, config) {
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return client.post('package/get', _objectSpread({}, params), _objectSpread({
                expect: _EXPECT,
                responsePath: 'package'
              }, config));
            case 2:
              return _context3.abrupt("return", _context3.sent);
            case 3:
            case "end":
              return _context3.stop();
          }
        }, _callee3);
      }));
      function get(_x5, _x6, _x7) {
        return _get.apply(this, arguments);
      }
      return get;
    }()
  }, {
    key: "list",
    value: function () {
      var _list = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(client, params, config) {
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return client.post('package/list', _objectSpread({}, params), _objectSpread({
                expect: _EXPECT_LIST
              }, config));
            case 2:
              return _context4.abrupt("return", _context4.sent);
            case 3:
            case "end":
              return _context4.stop();
          }
        }, _callee4);
      }));
      function list(_x8, _x9, _x10) {
        return _list.apply(this, arguments);
      }
      return list;
    }()
  }]);
  return Package;
}();
exports.Package = Package;
