// @steamship/client/build/main/lib/tag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tag = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _EXPECT = function _EXPECT(client, data) {
  return new Tag(client, data);
};
var _EXPECT_LIST = function _EXPECT_LIST(client, data) {
  return {
    tags: data.tags.map(function (x) {
      return _EXPECT(client, x);
    })
  };
};
var Tag = /*#__PURE__*/function () {
  function Tag(client, params) {
    (0, _classCallCheck2["default"])(this, Tag);
    (0, _defineProperty2["default"])(this, "id", void 0);
    (0, _defineProperty2["default"])(this, "fileId", void 0);
    (0, _defineProperty2["default"])(this, "blockId", void 0);
    (0, _defineProperty2["default"])(this, "workspaceId", void 0);
    (0, _defineProperty2["default"])(this, "startIdx", void 0);
    (0, _defineProperty2["default"])(this, "endIdx", void 0);
    (0, _defineProperty2["default"])(this, "kind", void 0);
    (0, _defineProperty2["default"])(this, "name", void 0);
    (0, _defineProperty2["default"])(this, "value", void 0);
    (0, _defineProperty2["default"])(this, "text", void 0);
    (0, _defineProperty2["default"])(this, "createdAt", void 0);
    (0, _defineProperty2["default"])(this, "updatedAt", void 0);
    (0, _defineProperty2["default"])(this, "client", void 0);
    this.client = client;
    this.id = params.id;
    this.fileId = params.fileId;
    this.blockId = params.blockId;
    this.workspaceId = params.workspaceId;
    this.startIdx = params.startIdx;
    this.endIdx = params.endIdx;
    this.kind = params.kind;
    this.name = params.name;
    this.value = params.value;
    this.text = params.text;
    this.createdAt = params.createdAt;
    this.updatedAt = params.updatedAt;
  }
  (0, _createClass2["default"])(Tag, [{
    key: "delete",
    value: function () {
      var _delete2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(config) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return this.client.post('tag/delete', {
                id: this.id
              }, _objectSpread({
                expect: _EXPECT,
                responsePath: 'tag'
              }, config));
            case 2:
              return _context.abrupt("return", _context.sent);
            case 3:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function _delete(_x) {
        return _delete2.apply(this, arguments);
      }
      return _delete;
    }()
  }], [{
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, params, config) {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return client.post('tag/create', params, _objectSpread(_objectSpread({}, config), {}, {
                expect: _EXPECT,
                responsePath: 'tag'
              }));
            case 2:
              return _context2.abrupt("return", _context2.sent);
            case 3:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      function create(_x2, _x3, _x4) {
        return _create.apply(this, arguments);
      }
      return create;
    }()
  }, {
    key: "query",
    value: function () {
      var _query = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, tagFilterQuery, config) {
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return client.post('tag/query', {
                tagFilterQuery: tagFilterQuery
              }, _objectSpread({
                expect: _EXPECT_LIST
              }, config));
            case 2:
              return _context3.abrupt("return", _context3.sent);
            case 3:
            case "end":
              return _context3.stop();
          }
        }, _callee3);
      }));
      function query(_x5, _x6, _x7) {
        return _query.apply(this, arguments);
      }
      return query;
    }()
  }]);
  return Tag;
}();
exports.Tag = Tag;
