// @steamship/client/build/main/lib/user.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.User = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _EXPECT = function _EXPECT(client, data) {
  if (data.user) {
    data = data.user;
  }
  return new User(client, data);
};
var User = /*#__PURE__*/function () {
  function User(client, params) {
    (0, _classCallCheck2["default"])(this, User);
    (0, _defineProperty2["default"])(this, "client", void 0);
    (0, _defineProperty2["default"])(this, "id", void 0);
    (0, _defineProperty2["default"])(this, "firstName", void 0);
    (0, _defineProperty2["default"])(this, "lastName", void 0);
    (0, _defineProperty2["default"])(this, "nickName", void 0);
    (0, _defineProperty2["default"])(this, "handle", void 0);
    (0, _defineProperty2["default"])(this, "plan", void 0);
    (0, _defineProperty2["default"])(this, "handleSet", void 0);
    (0, _defineProperty2["default"])(this, "profile", void 0);
    (0, _defineProperty2["default"])(this, "profilePhoto", void 0);
    this.client = client;
    this.id = params.id;
    this.firstName = params.firstName;
    this.lastName = params.lastName;
    this.nickName = params.nickName;
    this.handle = params.handle;
    this.plan = params.plan;
    this.handleSet = params.handleSet;
    this.profile = params.profile;
    this.profilePhoto = params.profilePhoto;
  }
  (0, _createClass2["default"])(User, [{
    key: "update",
    value: function () {
      var _update = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(params, config) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return this.client.post('account/update', params || {}, _objectSpread({
                expect: _EXPECT
              }, config));
            case 2:
              return _context.abrupt("return", _context.sent);
            case 3:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function update(_x, _x2) {
        return _update.apply(this, arguments);
      }
      return update;
    }()
  }], [{
    key: "current",
    value: function () {
      var _current = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, config) {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return client.get('account/current', {}, _objectSpread({
                expect: _EXPECT
              }, config));
            case 2:
              return _context2.abrupt("return", _context2.sent);
            case 3:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      function current(_x3, _x4) {
        return _current.apply(this, arguments);
      }
      return current;
    }()
  }]);
  return User;
}();
exports.User = User;
