// @steamship/cli/dist/commands/workspace/list.js
import { CliUx, Command } from '@oclif/core';
import { Workspace } from '@steamship/client';
import { getAuthedClient } from '../../helpers/get-client.js';
import { showTable } from '../../helpers/show-table.js';
function toList(d) {
    return [d.handle || '', d.description || '', d.createdAt || ''];
}
export default class List extends Command {
    static description = 'Lists your workspaces.';
    static examples = ['ship workspace:list'];
    static flags = {};
    static args = [];
    async run() {
        const client = await getAuthedClient();
        const request = await Workspace.list(client);
        if (!request.output) {
            CliUx.ux.error('Could not retrieve workspace list from Steamship.');
        }
        const workspaces = request.output?.workspaces;
        if (!workspaces || workspaces.length == 0) {
            CliUx.ux.info("You don't have any workspaces yet.");
            CliUx.ux.exit();
        }
        showTable({
            title: 'Workspaces',
            headers: ['Handle', 'Description', 'Created At'],
            data: workspaces.map(toList)
        });
    }
}
