// @steamship/cli/dist/commands/plugin/update.js
import { CliUx, Command, Flags } from '@oclif/core';
import { getAuthedClient } from '../../helpers/get-client.js';
import { getProjectConfigWizard } from '../../wizards/get-project-config.js';
import { loadReadmeFile } from "../../helpers/project-config.js";
import { Plugin } from "@steamship/client";
export default class Update extends Command {
    static description = 'Updates a Steamship Plugin in the registry';
    static examples = ['ship plugin:update'];
    static flags = {
        help: Flags.help(),
        version: Flags.version()
    };
    static args = [];
    async run() {
        const { args, flags } = await this.parse(Update);
        const manifest = await getProjectConfigWizard({
            checkName: true,
            checkHandle: true,
            checkVersion: true,
            checkType: true,
            checkEntrypoint: true,
            checkPluginParams: true,
            willDeploy: true
        });
        if (manifest.type != 'plugin') {
            CliUx.ux.error('The plugin:update command must be run.js from the root folder of a Steamship Plugin project.');
        }
        CliUx.ux.log(`Updating Plugin ${manifest.handle}`);
        const description = manifest.description;
        const readme = loadReadmeFile();
        const client = await getAuthedClient();
        const resp = await Plugin.get(client, { handle: manifest.handle });
        if ((!resp) || (!resp.output)) {
            CliUx.ux.error(`Unable to load plugin with handle ${manifest.handle}`);
        }
        const inst = resp.output;
        inst.readme = readme;
        inst.description = description;
        inst.profile = manifest;
        CliUx.ux.log(`Updating Plugin ${inst.handle}`);
        await inst.update();
        CliUx.ux.log(`Updated!`);
    }
}
