// @steamship/client/build/module/lib/shared/Configuration.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (it) return (it = it.call(o)).next.bind(it); if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, defineProperty = Object.defineProperty || function (obj, key, desc) { obj[key] = desc.value; }, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return defineProperty(generator, "_invoke", { value: makeInvokeMethod(innerFn, self, context) }), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == typeof value && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; defineProperty(this, "_invoke", { value: function value(method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); } }); } function makeInvokeMethod(innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; } function maybeInvokeDelegate(delegate, context) { var methodName = context.method, method = delegate.iterator[methodName]; if (undefined === method) return context.delegate = null, "throw" === methodName && delegate.iterator["return"] && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method) || "return" !== methodName && (context.method = "throw", context.arg = new TypeError("The iterator does not provide a '" + methodName + "' method")), ContinueSentinel; var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, defineProperty(Gp, "constructor", { value: GeneratorFunctionPrototype, configurable: !0 }), defineProperty(GeneratorFunctionPrototype, "constructor", { value: GeneratorFunction, configurable: !0 }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (val) { var object = Object(val), keys = []; for (var key in object) keys.push(key); return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, "catch": function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }
function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import { isNode } from '../utils.js';

/*
 * The goal of this file is to provide a credential-loading system
 * that is safe in BOTH the command line and the browser.
 */

export var DEFAULT_CONFIG = {
  apiBase: 'https://api.steamship.com/api/v1/',
  appBase: 'https://steamship.run/',
  webBase: 'https://app.steamship.com/'
};

// const defaultStagingCredentials = {
//   apiBase: "https://api.staging.steamship.com/api/v1",
//   appBase: "https://staging.steamship.com/api/v1"
// }

export var CONFIG_FILENAME = '.steamship.json';
var ConfigManager = /*#__PURE__*/function () {
  function ConfigManager() {
    _classCallCheck(this, ConfigManager);
    _defineProperty(this, "_readFile", undefined);
    _defineProperty(this, "_writeFile", undefined);
    _defineProperty(this, "_configSearchPath", []);
    _defineProperty(this, "_config", {});
  }
  _createClass(ConfigManager, [{
    key: "prepare",
    value: function () {
      var _prepare = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
        var fs, os, path;
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 3;
              return import('fs');
            case 3:
              fs = _context.sent;
              this._readFile = function (filePath) {
                if (!fs.existsSync(filePath)) {
                  return undefined;
                }
                return fs.readFileSync(filePath, 'utf8');
              };
              this._writeFile = function (filePath, data) {
                fs.writeFileSync(filePath, data);
              };
              _context.next = 11;
              break;
            case 8:
              _context.prev = 8;
              _context.t0 = _context["catch"](0);
              // Running from the browser
              this._readFile = function () {
                return undefined;
              };
            case 11:
              _context.prev = 11;
              if (!isNode()) {
                _context.next = 20;
                break;
              }
              _context.next = 15;
              return import('os');
            case 15:
              os = _context.sent;
              _context.next = 18;
              return import('path');
            case 18:
              path = _context.sent;
              // Try the home folder last.
              this._configSearchPath.push(path.join(os.homedir(), CONFIG_FILENAME));
            case 20:
              _context.next = 26;
              break;
            case 22:
              _context.prev = 22;
              _context.t1 = _context["catch"](11);
              console.log(_context.t1);
              console.log('Unable to load config from search path');
              // pass
            case 26:
            case "end":
              return _context.stop();
          }
        }, _callee, this, [[0, 8], [11, 22]]);
      }));
      function prepare() {
        return _prepare.apply(this, arguments);
      }
      return prepare;
    }()
  }, {
    key: "_read",
    value: function () {
      var _read2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(filePath, profile) {
        var str, json;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              if (filePath) {
                _context2.next = 2;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 2:
              if (isNode()) {
                _context2.next = 4;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 4:
              if (!(typeof this._readFile == 'undefined')) {
                _context2.next = 6;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 6:
              str = this._readFile(filePath);
              if (str) {
                _context2.next = 9;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 9:
              // Let the parsing error trickle up!
              // We want the user to know their config has a problem.
              json = JSON.parse(str);
              if (!profile) {
                _context2.next = 18;
                break;
              }
              if (json['profiles']) {
                _context2.next = 13;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 13:
              if (json['profiles'][profile]) {
                _context2.next = 15;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 15:
              return _context2.abrupt("return", json['profiles'][profile]);
            case 18:
              return _context2.abrupt("return", json);
            case 19:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _read(_x, _x2) {
        return _read2.apply(this, arguments);
      }
      return _read;
    }()
  }, {
    key: "clear",
    value: function clear() {
      this._config = {
        apiBase: DEFAULT_CONFIG.apiBase,
        appBase: DEFAULT_CONFIG.appBase,
        webBase: DEFAULT_CONFIG.webBase
      };
    }
  }, {
    key: "loadFromFile",
    value: function () {
      var _loadFromFile = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(filename, profile) {
        var config;
        return _regeneratorRuntime().wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              if (filename) {
                _context3.next = 2;
                break;
              }
              return _context3.abrupt("return", false);
            case 2:
              _context3.next = 4;
              return this._read(filename, profile);
            case 4:
              config = _context3.sent;
              if (config) {
                _context3.next = 7;
                break;
              }
              return _context3.abrupt("return", false);
            case 7:
              if (config.apiBase) {
                this._config.apiBase = config.apiBase;
              }
              if (config.appBase) {
                this._config.appBase = config.appBase;
              }
              if (config.webBase) {
                this._config.webBase = config.webBase;
              }
              if (config.apiKey) {
                this._config.apiKey = config.apiKey;
              }
              if (config.workspaceId) {
                this._config.workspaceId = config.workspaceId;
              }
              if (config.workspaceHandle) {
                this._config.workspaceHandle = config.workspaceHandle;
              }
              return _context3.abrupt("return", true);
            case 14:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function loadFromFile(_x3, _x4) {
        return _loadFromFile.apply(this, arguments);
      }
      return loadFromFile;
    }()
  }, {
    key: "tryAutofindingFiles",
    value: function () {
      var _tryAutofindingFiles = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee4(profile) {
        var _iterator, _step, configPath, found;
        return _regeneratorRuntime().wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _iterator = _createForOfIteratorHelperLoose(this._configSearchPath);
            case 1:
              if ((_step = _iterator()).done) {
                _context4.next = 10;
                break;
              }
              configPath = _step.value;
              _context4.next = 5;
              return this.loadFromFile(configPath, profile);
            case 5:
              found = _context4.sent;
              if (!found) {
                _context4.next = 8;
                break;
              }
              return _context4.abrupt("return", true);
            case 8:
              _context4.next = 1;
              break;
            case 10:
              return _context4.abrupt("return", false);
            case 11:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function tryAutofindingFiles(_x5) {
        return _tryAutofindingFiles.apply(this, arguments);
      }
      return tryAutofindingFiles;
    }()
  }, {
    key: "loadEnvVars",
    value: function () {
      var _loadEnvVars = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee5() {
        return _regeneratorRuntime().wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              if (typeof process != 'undefined' && typeof process.env != 'undefined') {
                try {
                  if (process.env['STEAMSHIP_API_BASE']) {
                    this._config.apiBase = process.env['STEAMSHIP_API_BASE'];
                  }
                  if (process.env['STEAMSHIP_APP_BASE']) {
                    this._config.appBase = process.env['STEAMSHIP_APP_BASE'];
                  }
                  if (process.env['STEAMSHIP_WEB_BASE']) {
                    this._config.webBase = process.env['STEAMSHIP_WEB_BASE'];
                  }
                  if (process.env['STEAMSHIP_API_KEY']) {
                    this._config.apiKey = process.env['STEAMSHIP_API_KEY'];
                  }
                  if (process.env['STEAMSHIP_WORKSPACE_ID']) {
                    this._config.workspaceId = process.env['STEAMSHIP_WORKSPACE_ID'];
                  }
                  if (process.env['STEAMSHIP_WORKSPACE_HANDLE']) {
                    this._config.workspaceHandle = process.env['STEAMSHIP_WORKSPACE_HANDLE'];
                  }
                } catch (_unused2) {
                  // pass
                }
              }
            case 1:
            case "end":
              return _context5.stop();
          }
        }, _callee5, this);
      }));
      function loadEnvVars() {
        return _loadEnvVars.apply(this, arguments);
      }
      return loadEnvVars;
    }()
  }, {
    key: "applyManualOverrides",
    value: function applyManualOverrides(params) {
      if (params != null && params.apiBase) {
        this._config.apiBase = params == null ? void 0 : params.apiBase;
      }
      if (params != null && params.appBase) {
        this._config.appBase = params == null ? void 0 : params.appBase;
      }
      if (params != null && params.webBase) {
        this._config.webBase = params == null ? void 0 : params.webBase;
      }
      if (params != null && params.apiKey) {
        this._config.apiKey = params == null ? void 0 : params.apiKey;
      }
      if (params != null && params.workspace) {
        this._config.workspaceHandle = params == null ? void 0 : params.workspace;
      }
    }
  }, {
    key: "load",
    value: function () {
      var _load = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee6(params) {
        var found;
        return _regeneratorRuntime().wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return this.prepare();
            case 2:
              this.clear();

              // First set the profile.
              if (typeof process != 'undefined' && typeof process.env != 'undefined') {
                if (process.env['STEAMSHIP_PROFILE']) {
                  this._config.profile = process.env['STEAMSHIP_PROFILE'];
                }
              }
              if (params != null && params.profile) {
                this._config.profile = params == null ? void 0 : params.profile;
              }

              // Then load from a file if provided
              if (!(params != null && params.configFile)) {
                _context6.next = 13;
                break;
              }
              _context6.next = 8;
              return this.loadFromFile(params == null ? void 0 : params.configFile, this._config.profile);
            case 8:
              found = _context6.sent;
              if (found) {
                _context6.next = 11;
                break;
              }
              throw "Configuration path provided but no configuration found at '" + (params == null ? void 0 : params.configFile) + "'";
            case 11:
              _context6.next = 15;
              break;
            case 13:
              _context6.next = 15;
              return this.tryAutofindingFiles(this._config.profile);
            case 15:
              _context6.next = 17;
              return this.loadEnvVars();
            case 17:
              // Apply manual overrides
              this.applyManualOverrides(params);

              // Fix the base
              if (this._config.apiBase) {
                if (this._config.apiBase[this._config.apiBase.length - 1] != '/') {
                  this._config.apiBase = this._config.apiBase + "/";
                }
              }
              if (this._config.appBase) {
                if (this._config.appBase[this._config.appBase.length - 1] != '/') {
                  this._config.appBase = this._config.appBase + "/";
                }
              }
              if (this._config.webBase) {
                if (this._config.webBase[this._config.webBase.length - 1] != '/') {
                  this._config.webBase = this._config.webBase + "/";
                }
              }
              return _context6.abrupt("return", this._config);
            case 22:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this);
      }));
      function load(_x6) {
        return _load.apply(this, arguments);
      }
      return load;
    }()
  }, {
    key: "modifyConfiguration",
    value: function () {
      var _modifyConfiguration = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee7(values, profile, configFile) {
        var os, path, str, newConfig;
        return _regeneratorRuntime().wrap(function _callee7$(_context7) {
          while (1) switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return this.prepare();
            case 2:
              if (!(!isNode() || typeof this._readFile == 'undefined' || typeof this._writeFile == 'undefined')) {
                _context7.next = 4;
                break;
              }
              throw 'Could not save configuration, no filesystem found.';
            case 4:
              _context7.next = 6;
              return import('os');
            case 6:
              os = _context7.sent;
              _context7.next = 9;
              return import('path');
            case 9:
              path = _context7.sent;
              if (!(typeof configFile === 'string' && configFile.trim() === '')) {
                _context7.next = 12;
                break;
              }
              throw 'An invalid save location was given.';
            case 12:
              if (!configFile) {
                configFile = path.join(os.homedir(), CONFIG_FILENAME);
              }
              str = this._readFile(configFile);
              newConfig = _extends({}, DEFAULT_CONFIG);
              if (str) {
                try {
                  newConfig = JSON.parse(str);
                } catch (e) {
                  // If there is an issue with the config, we just use the default
                }
              }

              // Now set the desired values. If there is a profile, only do this to that
              // profile. Otherwise, override the fields.
              if (profile) {
                if (!newConfig.profiles) {
                  newConfig.profiles = {}; // Add the profiles field if it does not exist
                }

                if (!Object.keys(newConfig.profiles).includes(profile)) {
                  newConfig.profiles[profile] = {
                    profile: profile
                  }; // Adds this specific profile
                }

                Object.keys(values).forEach(function (key) {
                  if (key === 'profiles') {
                    throw 'Cannot save a profiles list within a profile itself. A `profiles` modification is only allowed on the default config (i.e. leave `profile` blank)';
                  }
                  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                  // @ts-ignore
                  newConfig.profiles[profile][key] = values[key];
                });
              } else {
                Object.keys(values).forEach(function (key) {
                  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                  // @ts-ignore
                  newConfig[key] = values[key];
                });
              }

              // Finally, save the configuration
              this._writeFile(configFile, JSON.stringify(newConfig, undefined, 2));
            case 18:
            case "end":
              return _context7.stop();
          }
        }, _callee7, this);
      }));
      function modifyConfiguration(_x7, _x8, _x9) {
        return _modifyConfiguration.apply(this, arguments);
      }
      return modifyConfiguration;
    }()
  }]);
  return ConfigManager;
}();
export function loadConfiguration(_x10) {
  return _loadConfiguration.apply(this, arguments);
}

/**
 * Modifies or creates a configuration given the desired values. Allows
 * modifying the default profile, or modifying individual profiles by
 * providing a `profile` parameter. If no configFile is given, defaults
 * to the default location (home directory / .steamship.json).
 * Note: If the configFile already exists, the existing file is modified.
 *       If the file is corrupt or missing, a new file is created
 * @param values The values to save to the config or profile
 * @param profile (Optional) The profile to override. Defaults to the default
 *                profile.
 * @param configFile (Optional) The absolute location to the file to save
 *                   to. Defaults to (home directory / .steamship.json)
 */
function _loadConfiguration() {
  _loadConfiguration = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee8(params) {
    var configManager;
    return _regeneratorRuntime().wrap(function _callee8$(_context8) {
      while (1) switch (_context8.prev = _context8.next) {
        case 0:
          configManager = new ConfigManager();
          _context8.next = 3;
          return configManager.load(params);
        case 3:
          return _context8.abrupt("return", _context8.sent);
        case 4:
        case "end":
          return _context8.stop();
      }
    }, _callee8);
  }));
  return _loadConfiguration.apply(this, arguments);
}
export function saveConfiguration(_x11, _x12, _x13) {
  return _saveConfiguration.apply(this, arguments);
}
function _saveConfiguration() {
  _saveConfiguration = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee9(values, profile, configFile) {
    var configManager;
    return _regeneratorRuntime().wrap(function _callee9$(_context9) {
      while (1) switch (_context9.prev = _context9.next) {
        case 0:
          configManager = new ConfigManager();
          _context9.next = 3;
          return configManager.modifyConfiguration(values, profile, configFile);
        case 3:
        case "end":
          return _context9.stop();
      }
    }, _callee9);
  }));
  return _saveConfiguration.apply(this, arguments);
}
